Key refactoring

daschuer opened the pull request:
This is some refactoring taken form https://github.com/mixxxdj/mixxx/pull/11096 rebased from 2.3. to 2.4
I have decided to do this in an extra PR, to have finally a clear view on the functional changes in the follow up PRs

daschuer committed:
Remove faulty and unnecessary Keys::isValid()

daschuer committed:
Rename all text based key representation to KeyText

daschuer committed:
Rename getGlobalKeyText() to formatGlobalKey(), because it reformats the global key according to the preferences.

daschuer committed:
split makebasicKeysFrom() Text into a normalizing and a keep text version

daschuer committed:
Improve TrackRecord::updateGlobalKeyText()

Swiftb0y commented on src/track/keys.cpp:
```diff
@@ -29,11 +29,6 @@ void Keys::setSubVersion(const QString& subVersion) {
     m_subVersion = subVersion;
 }
 
-bool Keys::isValid() const {
-    return m_keyMap.global_key() != mixxx::track::io::key::INVALID ||
-            m_keyMap.global_key_text().length() > 0;
-}
```
what was the motivation for removing this?

daschuer replied:
This is function has tricked me, because It hides implementation details. In all cases one of these checks is sufficient.  
I think it is better to explicit look for global_key or global_key_text in the demanding code.

Swiftb0y replied:
Mhmm, but won't that cause inconsistency problem then when the `global_key` is valid but the `global_key_text` is empty? Or is that a state that is to be expected?

daschuer replied:
The other way round can happen. There might be a key text that can not be recognized by Mixxx. Than `global_key` invalid but this function considers the whole object as valid, even if there is 'bla bla bla' stored as text, which is clearly invalid for a human reader. Because of this ambiguities it is better to check explicit the single members as desired in the calling code.

Swiftb0y commented on src/track/keyfactory.cpp:
```diff
@@ -22,27 +22,34 @@ Keys KeyFactory::loadKeysFromByteArray(const QString& keysVersion,
+        mixxx::track::io::key::Source source) {
     KeyMap key_map;
     key_map.set_global_key(global_key);
+    QString global_key_text = KeyUtils::keyToString(global_key, KeyUtils::KeyNotation::Custom);
+    key_map.set_global_key_text(global_key_text.toStdString());
```
this will cause a bunch of more mutex locks in a function that seems cheap. Are you sure this won't cause any problems?

daschuer replied:
Which mutex do you mean? When allocating memory? 
Do you have an alternative version in mind?

Swiftb0y replied:
`KeyUtils::keyToString` locks a mutex when using `KeyNotation::Custom`.

daschuer replied:
> KeyUtils::keyToString locks a mutex when using KeyNotation::Custom.

Ah, I see. This has only a limited scope and protects only against concurrent editing in the preferences. 
We may consider to use a copy during editing and atomic swap to get around it. 

Here is helps to have always a consistent key_text which will become the leading source of date with the next PR.

Swiftb0y reviewed (commented):
A couple questions

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -1252,7 +1245,7 @@ bool setTrackAudioProperties(
             mixxx::audio::Bitrate(bitrate),
             mixxx::Duration::fromSeconds(duration));
-    return false;
+    return false; // shouldDirty
```
This was my attempt to document the meaning of this anonymous boolean which is otherwise hard to understand. 
This value is copied into the variable shouldDirty, found here: https://github.com/mixxxdj/mixxx/blob/94b94e4a350830fe3e13d88cc12e100c74480f78/src/library/dao/trackdao.cpp#L1517 
Do you have an idea to improve the situation?

Swiftb0y replied:
Mhmm, that makes assumptions about the calling code. In most cases where a boolean is too meaningless, you can always just introduce an enum with two members. Then you can deduce the meaning from the name of the return type and you can deduce the meaning of the value from their names. 
Naming is hard so I struggle to come up with a concrete suggestion myself. Though the problem primarily stems from the fact that I struggle to understand the purpose of `setTrackBeats`/`setTrackAudioProperties`/etc. Maybe you can think of something?

daschuer replied:
I was about to fix it, but than I realized that all the code is not required. 
This was only used in setTrackKey() to replace a missing source flag with USER.
Even if the database is from an earlier version of Mixxx the data is consistent and the track can imported every time again.

daschuer committed:
No longer set tracks dirty after reading form a legacy database.

This was only used in setTrackKey() to replace a missing source flag with USER.
Even if the database is from an earlier version of Mixxx the data is consistent and the track can imported every time again.

daschuer commented:
CI failure is unrelated. See: https://github.com/mixxxdj/mixxx/issues/11386

daschuer committed:
Merge remote-tracking branch 'upstream/2.4' into key_refactoring

Swiftb0y commented on src/library/dao/trackdao.cpp:
```diff
@@ -1510,7 +1472,6 @@ TrackPointer TrackDAO::getTrackById(TrackId trackId) const {
     // of the properties has finished.
 
     // For every column run its populator to fill the track in with the data.
-    bool shouldDirty = false;
```
I don't get the rationale for removing this.

daschuer replied:
This is no longer used . The rational is described here: https://github.com/mixxxdj/mixxx/commit/1fc3e7477fc66cb273afd13348c0cb77a34e0280

Swiftb0y replied:
>  No longer set tracks dirty after reading form a legacy database.
> 
> This was only used in setTrackKey() to replace a missing source flag with USER.
Even if the database is from an earlier version of Mixxx the data is consistent and the track can imported every time again.

But won't that mean the in the non-legacy database that the source flag will be set to an undefined value or missing?

daschuer replied:
The adopting of the USER state in RAM in this case remains. 
In the current released version, the missing keySource is set to USER in the Track Object and marked dirty which forces a write to track metadata and database. This writing is unnecessary, because Mixxx will adopt the USER state anyway again after restart.

Swiftb0y replied:
Ah I see. So the update is just postponed to next "flush" to the database?

daschuer replied:
Yes, when anything else dirties the track.

Swiftb0y commented on src/track/trackrecord.cpp:
```diff
@@ -45,14 +45,21 @@ bool TrackRecord::updateGlobalKey(
     }
-    setKeys(keys);
+
+    Keys newKeys = KeyFactory::makeBasicKeys(newKey, keySource);
```
won't this normalize user key data? eg shouldn't this be `makeBasicKeysKeepText`?

daschuer replied:
Yes. We can discuss about this. 

For now I have kept the original behaviour that user edits are checked and normalized to make sure the string written to the file metadata is useful for other tools. So in case the user enters a string that Mixxx understands, but violates the id3 format it is corrected. 

makeBasicKeysKeepText() is used to not touch a string that is already written into the file metadata. Checking and modifying it is not the busyness of Mixxx IMHO.

I can imagine a use case where the user want to manual edit a non standard string but there is a risk to mess the strings up. Without a special use case I prefere not to open it. 

What do you think? 

I think I will add a comment.

daschuer replied:
Oh I forget. This PR is only the refactoring PR without functional changes. The considerations above are useful for the follow up PR.

Swiftb0y replied:
Well the concern I had when I commented this initially was that the previous behavior was more relaxed and did not normalize. Since the method is called `TrackRecord::updateGlobalKeyText` I thought this code is not exclusive to user-edits but also gets run on import of pre-existing metadata which is why I thought that shouldn't be normalized. If this code is only run on user-edit within mixxx, normalizing would be the correct thing to do IMO.

daschuer replied:
I have renamed the function. It runs in both cases fixing this in the "read from file metadata" case is the topic of the follow up PR. The intention of this PR is not to change the behavior.

Swiftb0y replied:
I'm sorry. I've misunderstood the code. LGTM.

daschuer committed:
rename updateGlobalKeyNormalizeText(

Swiftb0y reviewed (approved):
LGTM
